### PR TITLE
Unicode documentation

### PR DESCRIPTION
**Pull request by mosteo**
I think I have now a clear understanding of the implications of `-gnatW8`. I've tried to document the practical issues to put the matter to rest. So this PR includes my findings in an AEP file for our internal record, and also a document intended to go into the website for people that come looking for information on that.

I've also added the option to disable `-gnatW8` just so if in some case it really doesn't make sense (not for the public index, but maybe in some legacy private codebase) you're not forced to disable the whole crate configuration system.

Next steps would be to implement the check for the index submissions and to fix the few crates that cause problems.

Besides you @Fabien-Chouteau, it would be great if other people familiar with the issue (@reznikmm, ...) can verify I'm not making some wrong assumptions or statements in these documents. The `doc/unicode.md` can probably be improved as I tend to err to the convoluted side.

### CODE REVIEW

File: src/alire/alire-utils-switches.ads [hunk 1]
```diff
@@ -52,6 +53,11 @@ is
                              Ada83, Ada95, Ada05, Ada12, Ada2022,
                              GNAT_Extensions);
 
+   type Unicode_Support_Kind is
+     (No_Support,  --  Default behavior of GNAT, no switches set (Latin1)
```
Comment (Fabien-Chouteau) [inline src/alire/alire-utils-switches.ads]: I don't think those names are meaningful enough.

File: doc/unicode.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    1. By default, Ada.Wide_Wide_Text_IO will use UTF-8 encoding if `-gnatW8` is in effect.
+    1. Most modern plain text formats expect UTF-8 encoding.
+    1. On Linux terminals this is the usual default.
+    1. On Windows 10:
+        1. You can enable UTF-8 once by running `chcp 65001` in your terminal.
+        1. You can enable UTF-8 in terminals system-wide.
+            - This will break programs assuming the default Windows 10 encoding.
+    1. On Windows 11, UTF-8 terminals are the default.
+1. You can use `GNAT.IO` or `Ada.Streams.Stream_IO` to output properly encoded strings, as they don't manipulate the bytes.
+1. You **cannot** use `Ada.Text_IO` to output UTF-8-encoded `String` variables, as Latin-1 encoding is expected.
```
Comment (Fabien-Chouteau) [inline doc/unicode.md:66]: That sounds scary since every Ada will ask you to use `Ada.Text_IO` to print strings.

**Review by Fabien-Chouteau — COMMENTED** [review]
The more I see about this issue the more I think this is not for Alire to fix. Something has to change either in the language or the compiler (maybe both).

File: doc/unicode.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+--  Without -gnatW8 but with an UTF-8 file it will fail as the sequence will be 6 bytes long.
+
+Euro : constant String := "€";
+--  Without -gnatW8 and a UTF-8-encoded file, this results in a 3-byte UTF-8 string.
+--  With -gnatW8 and a UTF-8-encoded file, this results in an error as '€' is outside of Character.
+```
+
+## Unicode in identifiers
+
+By default, GNAT only allows Latin-1 identifiers. To enable full Unicode identifiers you can use `-gnatiw`. This is generally not recommended, as that may greatly difficult others to work with your code. Alire does not enable full Unicode in identifiers by default.
```
Comment (reznikmm) [inline doc/unicode.md]: In presence of `-gnatW8` no other switches is required to have non-ASCII identifiers. So `-gnatiw` only required if there is no `-gnatW8`.

Comment (mosteo) [reply doc/unicode.md]: Right, this was a leftover.

File: doc/unicode.md (hunk 2 above)
Comment (mosteo) [reply doc/unicode.md:66]: Yeah, it's the ugliest aspect of this issue. Unfortunately this works "by chance" when the terminal is UTF-8, the file is UTF-8, but no `-gnatW8` is given, as TIO lets "fake" Latin-1 strings which are UTF-8 get by unmodified. If the source is properly Latin-1-encoded then the output of TIO is again wrong [meaning it won't display properly in UTF-8 terminals].

**Comment by mosteo** [comment]
> The more I see about this issue the more I think this is not for Alire to fix. Something has to change either in the language or the compiler (maybe both).

Can you get a feel about this from people in AdaCore? What are people doing in their codebases? Is this a usual thing in coding conventions? What I've seen in comp.lang.ada is that using `Ada.Text_IO` for Unicode has always been seen as a mistake and asking for trouble.

I don't think the language is going to change, it already admits full Unicode. It's more of a problem with legacy codebases and compiler defaults, unlikely to change either, I'd say.

We are in the crossfire of wanting code from multiple origins to always interact properly. Closed projects likely have policies about this and is a non-issue. That said, I tend to think we're in the proper spot to adopt this now and get done with it for good.

I will try to get more eyes/feedback on it.

File: src/alire/alire-utils-switches.ads (hunk 1 above)
Comment (mosteo) [reply src/alire/alire-utils-switches.ads]: What about `Source_Encoding`/`ASCII`/`Latin_1`/`UTF_8`? (ASCII/Latin_1 would be equivalent -- no `-gnatW8`)

File: doc/unicode.md (hunk 3 above)
Comment (mosteo) [reply doc/unicode.md]: Actually that was already removed in a later commit. Better if you check the changes as a whole instead of commit-by-commit.

**Comment by stcarrez** [comment]
This `-gnatW8` option breaks a simple Ada program which uses `Ada.Directories` and prints file names.
See https://github.com/alire-project/alire/discussions/1334#discussioncomment-5205705

Asking people not to use `Ada.Text_IO` is definitely a very bad idea.

Unless the compiler is fixed, we will have many issues if you set this option by default.

**Review by Fabien-Chouteau — APPROVED** [review]

**Comment by mosteo** [comment]
I will merge this one then, as I think it's better to have some rationale rather than just simply enabling `-gnatW8` silently, but improvements to the documentation are welcomed.

**Pull request merged**